# CDDR_Variants_filtering
First make the variants data ready...
* This is the data from the CDDR bulls sequencing project, aligned to ARS_UCD.v14 (variants called using bcftools) and lifted over to umd3 assembly (using a liftover chain generated by minmap2 alignment and sam2psl converter). 
* All the vcf files (chr1tochrX) are lifted over (using picard liftovervcf) and combined (using bcftools concat)
* Combined vcf variants are given ids (bcftools annotate setID)

```bash
# working directory
/mnt/nfs/nfs1/derek.bickhart/CDDR-project/vcfs/condensed-vcfs/liftover/annotated/filtration/
module load bcftools samtools
# Now to sort according to Chr and pos:
[kiranmayee.bakshy@assembler2 liftover]$ cat <(zcat combinedvcfs_id.gz | grep -B10000 -m1 ^#CHROM) <(bcftools view -H combinedvcfs_id.gz | sort -k1V,1 -k2n,2) | bgzip > combined.ids.sorted.vcf.gz
# Index the large vcf
tabix -p vcf combined.ids.sorted.vcf.gz
```

Now start filtering:
## Step 1: select variants (only snps) in the selected haplotype segment coordinates:
```bash
[kiranmayee.bakshy@assembler2 filtration]$ bcftools view -m2 -M2 -v snps /mnt/nfs/nfs1/derek.bickhart/CDDR-Project/vcfs/condensed_vcfs/liftover/combined.ids.sorted.vcf.gz -R seg_cord.bed -Oz -o f1.vcf.gz

#Get summary stats
[kiranmayee.bakshy@assembler2 filtration]$ bcftools stats f1.vcf.gz | grep -P "SN\t"
# SN    [2]id   [3]key  [4]value
SN      0       number of samples:      172
SN      0       number of records:      17161409
SN      0       number of no-ALTs:      0
SN      0       number of SNPs: 17161409
SN      0       number of MNPs: 0
SN      0       number of indels:       0
SN      0       number of others:       0
SN      0       number of multiallelic sites:   0
SN      0       number of multiallelic SNP sites:       0

#Remove duplicate entries by chr, pos, ref and alt:
[kiranmayee.bakshy@assembler2 filtration]$ module load htslib samtools bcftools
[kiranmayee.bakshy@assembler2 filtration]$ /home/kiranmayee.bakshy/vcflib/bin/vcfuniq f1.vcf.gz | bgzip > f2.vcf.gz
[kiranmayee.bakshy@assembler2 filtration]$ bcftools stats f2.vcf.gz | grep -P "SN\t"
# SN    [2]id   [3]key  [4]value
SN      0       number of samples:      172
SN      0       number of records:      17156523
SN      0       number of no-ALTs:      0
SN      0       number of SNPs: 17156523
SN      0       number of MNPs: 0
SN      0       number of indels:       0
SN      0       number of others:       0
SN      0       number of multiallelic sites:   0
SN      0       number of multiallelic SNP sites:       0
```

## Step 2: Select only homozygous variants:
```bash
# Select just homozygous variant genotypes 
[kiranmayee.bakshy@assembler2 filtration]$ ~/vcflib/bin/vcffilter -g "GT = 1/1" f2.vcf.gz | bgzip > f3v1.vcf.gz

It is taking a lot of time...

# Select variants only in 147 samples -> update the AC,AF and AN 
[kiranmayee.bakshy@assembler2 filtration]$ ~/vcflib/bin/vcfkeepsamples f3v1.vcf.gz  001HO02706 001HO05072 001HO06478 001HO07714 001HO08730 001HO09404 001HO09736 001HO10852 001HO10983 001HO11361 001HO11386 001HO11389 001HO11393 007HO04985 007HO05992 007HO06480 007HO07004 007HO07430 007HO07777 007HO07930 007HO07956 007HO08233 007HO08387 007HO08437 007HO08492 007HO08573 007HO08759 007HO08771 007HO09221 007HO09289 007HO09306 007HO09411 007HO09426 007HO09438 007HO09993 007HO10168 007HO10588 007HO10676-0676_qia_test 007HO10687 007HO11042 007HO11332 007HO11372 007HO11420 007HO11610 007HO11737 007HO11850 007HO11868 007HO11901 007HO12014 007HO12033 007HO12174 007HO12223 007HO12292 007HO12353 007HO12356 007HO12359 007HO13016 007HO13061 009HO02140 011HO04544 011HO04594 011HO04911 011HO04943 011HO07794 011HO07823 011HO08363 011HO08538 011HO08833 011HO09198 011HO09224 011HO09867 011HO10013 011HO10268 011HO10331 011HO11088 011HO11124 011HO11176 011HO11231 011HO11316 014HO04280 014HO04493 014HO05680 014HO05770 014HO05797 014HO06154 014HO06441-6441_qia_test 014HO06648 014HO07049 014HO07063 014HO07259 014HO07283 014HO07503 014HO07519 014HO07561 014HO07629 021HO01136 023HO00598 029HO10124 029HO10390 029HO10826 029HO11428 029HO11816 029HO11993 029HO13083 029HO13089 029HO13121 029HO13339 029HO13561 029HO14038 029HO14136 029HO14830-14830_qia_test 029HO16177 029HO16313 029HO16493 029HO16608 029HO16688 029HO16805 029HO16901 029HO16955 029HO17606 029HO17651 029HO17674 029HO17681 029HO17816 040HO02054 071HO01486 073HO02770 073HO02828 076HO00485 080HO01077 080HO01102 094HO14819 094HO17755 200HO00337 200HO00402 200HO01143 200HO01954 200HO02353 200HO02504 200HO03453 200HO03588 200HO04518 200HO05649 200HO06150 200HO07375 200HO09805 250HO13232 | ~/vcflib/bin/vcffixup - | ~/vcflib/bin/vcffilter -f "AC > 0" | bgzip > f3v2.vcf.gz

 # Now select all the variants with AN=294 (i.e. no. of samples, 147 times 2)
[kiranmayee.bakshy@assembler2 filtration]$ bcftools view -i 'AN==294' -Oz -o f3v3.vcf.gz f3v2.vcf.gz
# Generate summary stats
[kiranmayee.bakshy@assembler2 filtration]$ bcftools stats -v -s - -F /mnt/nfs/nfs2/Genomes/umd3_kary_unmask_ngap.fa --af-bins '(0.05,0.1,0.5,1)' f3v3.vcf.gz > f3v3.stats
[kiranmayee.bakshy@assembler2 filtration]$ plot-vcfstats -r -p f3v3 f3v3.stats
```

## Step 3: Selecting variants within the CDDR sequenced samples
```bash
# Excluding the variants with > 10% missing genotypes 
[kiranmayee.bakshy@assembler2 filtration]$ bcftools view -S NAAB_samples_overlap.txt f2.vcf.gz | bcftools filter -i "F_MISSING<0.1" | bgzip -c > f3.vcf.gz && tabix f3.vcf.gz

# Selecting variants with high quality, QUAL >= 999, conform to HWEP,  0.05 > MAF < 0.50 and no read bias (RPB > 0.1)
[kiranmayee.bakshy@assembler2 filtration]$ bcftools filter -i "QUAL>=999 & MAF>0.05 & MAF<0.5 & RPB>0.1" f3.vcf.gz | bgzip -c > f5.vcf.gz && tabix f5.vcf.gz

# Merge above 2 steps 
[kiranmayee.bakshy@assembler2 filtration]$ bcftools view -S NAAB_samples_overlap.txt f2.vcf.gz | bcftools filter -i "F_MISSING<0.1 & QUAL>=999 & MAF>0.05 & MAF<0.5 & RPB>0.90" | bgzip -c > f4.vcf.gz && tabix f4.vcf.gz
```
## Step 4: Selecting homozygous variants
```bash
# method to call the sites that are all homozygous Variants (1/1, 2/2 genotype)
GenomeAnalysisTK -T SelectVariants -R /mnt/nfs/nfs1/derek.bickhart/CDDR-Project/vcfs/condensed_vcfs/liftover/test_files/umd3_kary_unmask_ngap.fa -V f2.vcf.gz -o f_gt_1.vcf -select 'vc.getHomVarCount() == 6'
## outcome: no sites will have ambigous (./.) and/or homRef (0/0). All other homVar are allowed

# select vcf sites where no sample has hetVar (no Gt = 0/1, 0/2), and at least one sample is HomVar (GT = 1/1)
GenomeAnalysisTK -T SelectVariants -R /mnt/nfs/nfs1/derek.bickhart/CDDR-Project/vcfs/condensed_vcfs/liftover/test_files/umd3_kary_unmask_ngap.fa -V f2.vcf.gz -o f_gt_2.vcf -select 'vc.getHetCount() == 0 && vc.getHomVarCount() >= 1'
```



```bash
Try these later...
# site where at least 1 sample is homVar (GT = 1/1 and/or 2/2 etc.)
GenomeAnalysisTK -T SelectVariants -R /mnt/nfs/nfs1/derek.bickhart/CDDR-Project/vcfs/condensed_vcfs/liftover/test_files/umd3_kary_unmask_ngap.fa -V f2.vcf.gz -o f_gt_1.vcf -select 'vc.getHomVarCount() >= 1'

# at least 1 sample is homVar (1/1, 2/2 or 3/3)
GenomeAnalysisTK -T SelectVariants -R /mnt/nfs/nfs1/derek.bickhart/CDDR-Project/vcfs/condensed_vcfs/liftover/test_files/umd3_kary_unmask_ngap.fa -V f2.vcf.gz -o f_gt_1.vcf -select 'vc.getHomVarCount() != 0'


```





