# CDDR_Variants_filtering
First make the variants data ready...
* This is the data from the CDDR bulls sequencing project, aligned to ARS_UCD.v14 (variants called using bcftools) and lifted over to umd3 assembly (using a liftover chain generated by minmap2 alignment and sam2psl converter). 
* All the vcf files (chr1tochrX) are lifted over (using picard liftovervcf) and combined (using bcftools concat)
* Combined vcf variants are given ids (bcftools annotate setID)

```bash
# working directory
/mnt/nfs/nfs1/derek.bickhart/CDDR-project/vcfs/condensed-vcfs/liftover/annotated/filtration/
module load bcftools samtools
# Now to sort according to Chr and pos:
[kiranmayee.bakshy@assembler2 liftover]$ cat <(zcat combinedvcfs_id.gz | grep -B10000 -m1 ^#CHROM) <(bcftools view -H combinedvcfs_id.gz | sort -k1V,1 -k2n,2) | bgzip > combined.ids.sorted.vcf.gz
# Index the large vcf
tabix -p vcf combined.ids.sorted.vcf.gz
```

Now start filtering:
## Step 1: select variants (only snps) in the selected haplotype segment coordinates:
```bash
[kiranmayee.bakshy@assembler2 filtration]$ bcftools view -m2 -M2 -v snps /mnt/nfs/nfs1/derek.bickhart/CDDR-Project/vcfs/condensed_vcfs/liftover/combined.ids.sorted.vcf.gz -R seg_cord.bed -Oz -o f1.vcf.gz

#Get summary stats
[kiranmayee.bakshy@assembler2 filtration]$ bcftools stats f1.vcf.gz | grep -P "SN\t"
# SN    [2]id   [3]key  [4]value
SN      0       number of samples:      172
SN      0       number of records:      17161409
SN      0       number of no-ALTs:      0
SN      0       number of SNPs: 17161409
SN      0       number of MNPs: 0
SN      0       number of indels:       0
SN      0       number of others:       0
SN      0       number of multiallelic sites:   0
SN      0       number of multiallelic SNP sites:       0

#Remove duplicate entries by chr, pos, ref and alt:
[kiranmayee.bakshy@assembler2 filtration]$ module load htslib samtools bcftools
[kiranmayee.bakshy@assembler2 filtration]$ /home/kiranmayee.bakshy/vcflib/bin/vcfuniq f1.vcf.gz | bgzip > f2.vcf.gz
[kiranmayee.bakshy@assembler2 filtration]$ bcftools stats f1.vcf.gz | grep -P "SN\t"






